pkg/resource/workspace/hooks.go

```diff
@@ -0,0 +1,220 @@
+		err = rm.updateWorkspaceAlias(ctx, desired)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &resource{ko}, nil
+}
+
+// updateWorkspaceTags uses TagResource and UntagResource to add, remove and update
+// workspace tags.
+func (rm *resourceManager) updateWorkspaceTags(
+	ctx context.Context,
+	latest *resource,
+	desired *resource,
+) error {
+	var err error
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.updateWorkspaceTags")
+	defer exit(err)
+
+	added, removed, updated := compareMaps(latest.ko.Spec.Tags, desired.ko.Spec.Tags)
+
+	if len(removed) > 0 {
+		removeTags := []*string{}
+		for i := range removed {
+			removeTags = append(removeTags, &removed[i])
+		}
+
+		input := &svcsdk.UntagResourceInput{
+			ResourceArn: (*string)(desired.ko.Status.ACKResourceMetadata.ARN),
+			TagKeys:     removeTags,
+		}
+
+		_, err = rm.sdkapi.UntagResourceWithContext(ctx, input)
+		rm.metrics.RecordAPICall("UPDATE", "UntagResource", err)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(updated)+len(added) > 0 {
+		addedTags := map[string]*string{}
+		for k, v := range added {
+			addedTags[k] = v
+		}
+		for k, v := range updated {
+			addedTags[k] = v
+		}
+
+		input := &svcsdk.TagResourceInput{
+			ResourceArn: (*string)(desired.ko.Status.ACKResourceMetadata.ARN),
+			Tags:        addedTags,
+		}
+		_, err = rm.sdkapi.TagResourceWithContext(ctx, input)
+		rm.metrics.RecordAPICall("UPDATE", "TagResource", err)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+
+}
+
+// updateWorkspaceAlias calls updateWorkspaceAlias to update a specific workspace
+// alias.
+func (rm *resourceManager) updateWorkspaceAlias(
+	ctx context.Context,
+	desired *resource,
+) error {
+	var err error
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.updateWorkspaceAlias")
+	defer exit(err)
+
+	input := &svcsdk.UpdateWorkspaceAliasInput{
+		WorkspaceId: desired.ko.Status.WorkspaceID,
+		Alias:       desired.ko.Spec.Alias,
+	}
+
+	_, err = rm.sdkapi.UpdateWorkspaceAliasWithContext(ctx, input)
+	rm.metrics.RecordAPICall("UPDATE", "UpdateWorkspaceAlias", err)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// compareMaps compares two string to string maps and returns three outputs: a
+// map of the new key/values observed, a list of the keys of the removed values
+// and a map containing the updated keys and their new values.
+func compareMaps(
+	a map[string]*string,
+	b map[string]*string,
+) (added map[string]*string, removed []string, updated map[string]*string) {
+	added = map[string]*string{}
+	updated = map[string]*string{}
+	visited := make(map[string]bool, len(a))
+	for keyA, valueA := range a {
+		valueB, found := b[keyA]
+		if !found {
+			removed = append(removed, keyA)
+			continue
+		}
+		if *valueA != *valueB {
+			updated[keyA] = valueB
+		}
+		visited[keyA] = true
+	}
+	for keyB, valueB := range b {
+		_, found := a[keyB]
+		if !found {
+			added[keyB] = valueB
+		}
+	}
+	return
+}
```

Comment:
I think you may have been copying this from elsewhere (Lambda controller maybe?) but you don't need to return three things here. Just a slice of strings of keys to delete and then a map of the added/updated k/v pairs.